Obstacles

Arsen-Lovecraft opened the pull request.

Arsen-Lovecraft committed:
init classes

Arsen-Lovecraft committed:
Merge branch 'main' into Obstacles

Arsen-Lovecraft committed:
Add code to obstacles (commit with bugs)

Add action "Interact"
Add binding in charact to interact function
Add multcast delegate on interact
Add folder for obstacles
Add code to obstacles but It's buggish and don't work as intended
Some formatting

Arsen-Lovecraft committed:
Add enemy obstacle(interactable) and obstacle

Comment: Now it's not very frindlt for blueprint, need some refactoring later.

fonkvan committed:
Merge branch 'main' into Obstacles

fonkvan committed:
Combined binaries

fonkvan commented:
Changes required:
-There is a bug with destroying enemies as you are not unbinding the function properly (should be "OnPlayerKilled" not "Destroy", or you can change Remove to RemoveDynamic and use &AEnemyObstacle::OnPlayerKilled in place of "destroy)
-You're doing some performance heavy stuff here by using IsA, which we can avoid by doing this:
```
if(ARailCharacter* RailCharacter = Cast<ARailCharacter>(OtherActor))
{
	RailCharacter->OnPlayerInteracted.AddDynamic(this, &AEnemyObstacle::OnPlayerKilled);
}
```
you can do the same in the onplayerendoverlap as well. This works does pretty much the same you had already done, and Cast already calls IsA (which is an expensive call), so the way you were doing it (by doing IsA and Cast) is really expensive. Just use cast only. You can also use Cast<To, From> instead of the cast I used if you prefer, but definitely change it to just a cast and don't use IsA


Change suggestions:
-Input_Interact should use ETriggerEvent::Triggered, not ETriggerEvent::Completed
-Interact function in player doesn't need to take FInputActionValue, as we don't use it
-Don't need to validate controller in Interact function (not using it)
-I would argue that the function name OnPlayerKilled is a little confusing. At first I thought it meant that the player was being killed, but really it's that the player is killing this enemy. A different name may make it clear what is actually happen (maybe OnInteract, OnPlayerInteracted, etc.)
-Personally, I also think you should use brackets for if statements, even if it's only one line. It's good practice to code this way in c++

Arsen-Lovecraft committed:
F

Made binding in EnemyObstacle.cpp consitent
Optimize cast in EnemyObstacle.cpp
Change enim of trigger event
Delete unused functionality

fonkvan reviewed (approved):
Thanks!
